Merge 'cardano-wallet-launcher' & 'cardano-wallet' together

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#368 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have introduced a new 'launch' command to do what `cardano-wallet-launcher` used to do.
- [x] I have removed the 'cardano-wallet-launcher' executable and replaced occurences of it in the tests by `cardano-wallet launch`
- [x] I have renamed `cardano-wallet server` to `cardano-wallet serve` as discussed briefly on slack
- [x] I have enabled `--port` and `--random-port` options for the `launch` command
- [x] I have slightly adjusted launcher code to use the new logging infrastructure

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

rvl committed:
Add DaedalusIPC module

This takes the cardano-sl/cardano-shell NodeIPC code and splits the
general nodejs child_process IPC protocol out from the
Daedalus/Cardano specific protocol, improves the exception handling,
changes the logging, and just makes it simpler.

rvl committed:
Integrate Daedalus IPC into CLI server start

rvl committed:
tests: Add nodejs child_process spawner

rvl committed:
tests: Add launcher --state-dir integration test

rvl committed:
tests: Add DaedalusIPC test

rvl committed:
DaedalusIPC: Remove Ping/Pong messages

rvl committed:
Regenerate nix

rvl committed:
DaedalusIPC: Remove Cabal library dependency (yuck)

rvl committed:
Fix rebase merge mistakes

rvl committed:
Address review comments

rvl committed:
DaedalusIPC: Don't use generic aeson encoding - too fragile

Removing a type constructor can change the shape of the JSON.

KtorZ committed:
merge cardano-wallet-launcher into cardano-wallet as a 'launch' command

KtorZ committed:
remove outdated comments in cardano-wallet executable

KtorZ committed:
nuke 'cardano-wallet-launcher' executable

KtorZ committed:
re-enable logging using tracer instead of 'say' in cardano-wallet executable

KtorZ committed:
enable port selection for 'cardano-wallet launch'

KtorZ committed:
add a few more examples of command for 'cardano-wallet'

KtorZ committed:
re-generate nix machinery :| ...

KtorZ commented on exe/wallet/Main.hs:
```diff
@@ -153,20 +172,26 @@ Usage:
-  --network <STRING>          testnet or mainnet [default: testnet]
-  --database <FILE>           use this file for storing wallet state
   --state <STRING>            address state: either used or unused
+  --state-dir <DIR>           write wallet state (blockchain and database) to this directory
```
I've sorted all the options alphabetically.

KtorZ commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Mnemonics.hs:
```diff
@@ -36,24 +36,6 @@ spec = do
-                out `shouldContain` "cardano-wallet-launcher"
-                c `shouldBe` ExitSuccess
-        forM_ ["-h", "--help"] test
-
```
The same tests already exist for `cardano-wallet`

KtorZ committed:
reduce usage of 'fixtureWallet' in the integration tests

There's a non-significant cost with the use of 'fixtureWallet' since it needs to restore a funded wallet
and wait until restoration. We also only have a limited number of those wallets, so it's better to keep
them for when we actually really need funded wallets. When testing parsing errors or wrong headers, there's
no need for demanding a wallet with funds

piotr-iohk commented on exe/wallet/Main.hs:
```diff
@@ -153,20 +172,26 @@ Usage:
+  cardano-wallet launch --network mainnet --random-port --state-dir .state-dir
+
+  # Start only a wallet server and connect to an already existing backend service
+  cardano-wallet serve --bridge-port 8080
```
It might be a bit confusing if someone does not have a context (i.e. is `wallet backend` different than `wallet server`) I think I'd rephrase this so it is more clear, e.g.:

```
  # Launch and monitor a wallet server and its associated backend service
  cardano-wallet launch --network mainnet --random-port --state-dir .state-dir

  # Start only a wallet server and connect to an already existing backend service
  cardano-wallet serve --bridge-port 8080
```

Anviking reviewed (commented):
lgtm based on a moderately quick look

KtorZ committed:
change 'cardano-wallet-launcher' to 'cardano-wallet launch'

KtorZ committed:
oversight -> use stack exec in daedalus mock instead of 'cardano-wallet-launcher'

KtorZ committed:
fix CLI/Server launcher tests not waiting long enough for the launcher to be up

KtorZ committed:
remove invalid amount '0', we've opened a bug about it

And right now, it's making the test fail because we do need a funded wallet to make this test. So it'll be best
to re-introduce it as part of a separate scenario, that we only run for the bridge
